balancing/balance_breaker.go

```diff
@@ -0,0 +1,615 @@
+}
+
+func (series *dataSeries) Add(value float64, dateTime time.Time) {
+	series.mx.Lock()
+	defer series.mx.Unlock()
+	series.data = append(series.data, &timeValue{dateTime, value})
+}
+
+func (series *dataSeries) ValueRangeFun(timeStart, timeEnd time.Time, fun func(*timeValue)) []float64 {
+	dataRange := []float64{}
+	for _, timeVal := range series.data {
+		if (timeStart == timeVal.date || timeStart.Before(timeVal.date)) && timeEnd.After(timeVal.date) {
+			fun(timeVal)
+		}
+	}
+	return dataRange
+}
+
+func (series *dataSeries) ValueRange(timeStart, timeEnd time.Time) []float64 {
+	dataRange := []float64{}
+	series.ValueRangeFun(timeStart, timeEnd, func(value *timeValue) {
+		dataRange = append(dataRange, value.value)
+	})
+	return dataRange
+}
+
+type timeValue struct {
+	date  time.Time
+	value float64
+}
+
+func newTimeHistogram(retention, resolution time.Duration, now func() time.Time) *histogram {
+	return &histogram{
+		t0:         now(),
+		resolution: resolution,
+		retention:  retention,
+		now:        now,
+		mx:         sync.Mutex{},
+	}
+}
+
+type histogram struct {
+	t0         time.Time
+	retention  time.Duration
+	resolution time.Duration
+	data       []*dataSeries
+	now        func() time.Time
+	mx         sync.Mutex
+}
+
+func (h *histogram) pickSeries(at time.Time) *dataSeries {
+	h.mx.Lock()
+	defer h.mx.Unlock()
+	idx := h.index(at)
+	if idx < 0 {
+		return nil
+	}
+	if idx >= h.cellsCount() || idx >= len(h.data) {
+		h.unshiftData(at)
+		idx = h.index(at)
+	}
+	return h.data[idx]
+}
+
+func (h *histogram) pickLastSeries(period time.Duration) []*dataSeries {
+	h.mx.Lock()
+	defer h.mx.Unlock()
+	if period > h.retention {
+		period = h.retention
+	}
+	h.unshiftData(h.now())
+	cellsNumber := math.Ceil(float64(period)/float64(h.resolution)) + 1
+	now := h.now()
+	stop := h.index(now) + 1
+	start := int(math.Max(float64(stop-int(cellsNumber)), 0))
+	return h.data[start:stop]
+}
+
+func (h *histogram) index(now time.Time) int {
+	sinceStart := float64(now.Sub(h.t0))
+	idx := math.Floor(sinceStart / float64(h.resolution))
+	return int(idx)
+}
+
+func (h *histogram) cellsCount() int {
+	return int(math.Ceil(float64(h.retention)/float64(h.resolution))) + 1
+}
+
+func (h *histogram) growSeries() {
+	for len(h.data) < h.cellsCount() {
+		h.data = append(h.data, &dataSeries{mx: sync.Mutex{}})
+	}
+}
+
+func (h *histogram) unshiftData(now time.Time) {
+	idx := h.index(now)
+	shiftSize := idx - len(h.data) + 1
+
+	if shiftSize > 0 && shiftSize < len(h.data) {
+		h.t0 = h.t0.Add(time.Duration(shiftSize) * h.resolution)
+		h.data = h.data[shiftSize:]
+	}
+	if shiftSize > 0 && len(h.data) > 0 && shiftSize >= len(h.data) {
+		h.t0 = now
+		h.data = []*dataSeries{}
+	}
+	h.growSeries()
+}
+func (h *histogram) shiftData(delta time.Duration) {
+	newT0 := h.t0.Add(delta)
+	if newT0.After(h.now()) {
+		return
+	}
+	h.t0 = newT0
+	for _, series := range h.data {
+		for _, value := range series.data {
+			value.date = value.date.Add(delta)
```

Comment:
This loop seems costly. We could instead accumulate the delta inside `histogram` and later, when someone calls `ValueRangeFun` just apply to delta to `timeStart` and `timeEnd` once at the beginning of the method